crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1755,11 +1756,17 @@ impl<'db> FunctionType<'db> {
     }
 
     /// inferred return type for this function
-    pub fn return_type(&self, db: &'db dyn Db) -> Type<'db> {
-        let definition = self.definition(db);
-        let DefinitionKind::Function(function_stmt_node) = definition.kind(db) else {
-            panic!("Function type definition must have `DefinitionKind::Function`")
-        };
+    ///
+    /// ## Why is this a salsa query?
+    ///
+    /// This is a salsa query to short-circuit the invalidation
+    /// when the function's AST node changes.
+    ///
+    /// Where this not a salsa query, then the calling query
+    /// would depend on the function's AST and rerun for every change in that file.
+    pub fn return_ty(self, db: &'db dyn Db) -> Type<'db> {
+        let scope = self.body_scope(db);
+        let function_stmt_node = scope.node(db).expect_function();
```

Comment:
The issue without this being a salsa query is that the calling-query suddenly depends on the function's ast (and, therefore, the file). 

In the case of our benchmark: `_parser.py` calls `match_to_datetime` which is defined in `_re.py`. Now, it shouldn't be necessary to recheck `_parser.py` because no type in `_re.py` changed. However, Salsa had to recheck every callsite because the `return_type` query accessed the `definition.kind(db)` which is the AST node of the `match_to_datetime` and its AST has changed (not really, but we use `no_eq`...)